<a href="https://colab.research.google.com/github/MatiasMoreno707/lab09-lp/blob/dev/lab09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Cree una red neuronal para clasificar a un paciente como canceroso o no canceroso de tal forma que obtenga el mayor valor de accuracy. Para ello, debe realizar el preprocesamiento de la información e ir ajustando la red neuronal con los parámetros requeridos como la cantidad de capas ocultas y neuronas por cada capa oculta, el número de iteraciones, la función de activación, el método para minimizar la función de error, entre otros.

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('cancer.csv')
df.head()

,Unnamed: 0,ClumpThick,UnifCellSize,UnifCellShape,MargAdh,SingEpiCellSize,BareNuclei,BlandChromatin,NormalNucleoli,Mitoses,Class
0,0,5,1,1,1,2,1,3,1,1,0
1,1,5,4,4,5,7,10,3,2,1,0
2,2,3,1,1,1,2,2,3,1,1,0
3,3,6,8,8,1,3,4,3,7,1,0
4,4,4,1,1,3,2,1,3,1,1,0


In [ ]:
def eliminar_columna(df, nombre_columna):
    """
    Elimina una columna específica de un DataFrame de pandas.

    Parámetros:
    -----------
    df : pandas.DataFrame
        El DataFrame del cual se eliminará la columna
    nombre_columna : str
        Nombre de la columna a eliminar

    Retorna:
    --------
    pandas.DataFrame
        DataFrame con la columna eliminada (None si la columna no existe)
    """
    try:
        df.drop(nombre_columna, axis=1, inplace=True)
        return df
    except KeyError:
        print(f"Advertencia: La columna '{nombre_columna}' no existe en el DataFrame")
        return None

In [ ]:
eliminar_columna(df, 'Unnamed: 0')

,ClumpThick,UnifCellSize,UnifCellShape,MargAdh,SingEpiCellSize,BareNuclei,BlandChromatin,NormalNucleoli,Mitoses,Class
0,5,1,1,1,2,1,3,1,1,0
1,5,4,4,5,7,10,3,2,1,0
2,3,1,1,1,2,2,3,1,1,0
3,6,8,8,1,3,4,3,7,1,0
4,4,1,1,3,2,1,3,1,1,0
...,...,...,...,...,...,...,...,...,...,...
694,3,1,1,1,3,2,1,1,1,0
695,2,1,1,1,2,1,1,1,1,0
696,5,10,10,3,7,3,8,10,2,1
697,4,8,6,4,3,4,10,6,1,1


In [ ]:
def porcentaje_vacios(df):
    """
    Muestra el porcentaje de valores vacíos para cada columna en un DataFrame.

    Parámetros:
    -----------
    df : pandas.DataFrame
        DataFrame a analizar

    Retorna:
    --------
    pandas.Series
        Serie con el porcentaje de valores vacíos por columna, ordenada descendente
    """
    return (df.isnull().mean() * 100).sort_values(ascending=False)

In [ ]:
porcentaje_vacios(df)

,0
ClumpThick,0.0
UnifCellSize,0.0
UnifCellShape,0.0
MargAdh,0.0
SingEpiCellSize,0.0
BareNuclei,0.0
BlandChromatin,0.0
NormalNucleoli,0.0
Mitoses,0.0
Class,0.0


In [ ]:
df.value_counts("Class")

,count
Class,
0,458
1,241


In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


In [ ]:
# Cargar dataset
df = load_breast_cancer()
X, y = df.data, df.target

# Dividir en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [ ]:
modelo_1 = MLPClassifier(hidden_layer_sizes=(30, 30),   # Dos capas ocultas con 30 neuronas cada una
                    activation='relu',             # Función de activación
                    solver='adam',                 # Método de optimización
                    max_iter=1000,                 # Iteraciones
                    random_state=42)

modelo_1.fit(X_train, y_train)


MLPClassifier(hidden_layer_sizes=(30, 30), max_iter=1000, random_state=42)

In [ ]:
y_pred = modelo_1.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nMatriz de confusión:\n", confusion_matrix(y_test, y_pred))
print("\nReporte de clasificación:\n", classification_report(y_test, y_pred))


Accuracy: 0.9649122807017544

Matriz de confusión:
 [[41  2]
 [ 2 69]]

Reporte de clasificación:
               precision    recall  f1-score   support

           0       0.95      0.95      0.95        43
           1       0.97      0.97      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.96      0.96       114
weighted avg       0.96      0.96      0.96       114



In [ ]:
modelo_2 = MLPClassifier(
    hidden_layer_sizes=(100,),
    activation='relu',
    max_iter=1000,
    solver='adam',
    alpha=0.0001,
    random_state=42
)

modelo_2.fit(X_train, y_train)

MLPClassifier(max_iter=1000, random_state=42)

In [ ]:
y_pred = modelo_2.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nMatriz de confusión:\n", confusion_matrix(y_test, y_pred))
print("\nReporte de clasificación:\n", classification_report(y_test, y_pred))

Accuracy: 0.9736842105263158

Matriz de confusión:
 [[41  2]
 [ 1 70]]

Reporte de clasificación:
               precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



| Métrica                 | **modelo_1**           | **modelo_2**                 |
|-------------------------|------------------------|------------------------------|
| **Accuracy**            | 0.9649                 | **0.9737** ✅                |
| **Precision (clase 0)** | 0.95                   | **0.98** ✅                  |
| **Recall (clase 0)**    | 0.95                   | 0.95                         |
| **Precision (clase 1)** | 0.97                   | 0.97                         |
| **Recall (clase 1)**    | 0.97                   | **0.99** ✅                  |
| **Matriz de confusión** | `[[41, 2], [2, 69]]`   | **`[[41, 2], [1, 70]]`** ✅ |
